src/compiler/resolutionCache.ts

```diff
@@ -449,7 +455,7 @@ namespace ts {
                 // Type references instead supply a `containingSourceFileMode` and a non-string entry which contains
                 // a default file mode override if applicable.
                 const mode = !isString(entry) ? getModeForFileReference(entry, containingSourceFileMode) :
-                    containingSourceFile ? getModeForResolutionAtIndex(containingSourceFile, i) : undefined;
+                    containingSourceFile ? getModeForResolutionAtIndex(containingSourceFile, partialResolutionInfo?.namesIndex[i] ?? i) : undefined;

```

Comment:
`partialResolutionInfo?.namesIndex`, in the one place its set, is the list of indices for unresolved names, but the list of `names` we have here is not guaranteed (in the context of the public API) to be that same list - it could easily be the whole list of all names, at which point this could easily be an out-of-bounds (or otherwise incorrect) index. That's not really that different from before, where this code largely assumed the `names` list was always the whole list of names. This feels like it can be wrong in almost the same way it could have been before in terms of public API usage, it just patches one internal callsite ([here](https://github.com/microsoft/TypeScript/pull/49738/files#diff-dbfa0e7f632eca196a5ee0e08e00743c46c03799e2704206cfccc065bb6f9e3aR1624)) for consistency. So I'm not sure this _really_ fixes the linked issue, so much as patches one potential exhibition of it.  By virtue of how much all this logic uses indices, it's hideously unsafe unless some unenforced invariants are upheld. Really, I think this logic needs a bit of a rework in terms of how cached resolutions and requested resolution names are related - these cache indexes are not great. This is sort-of a breaking API change as-is (it is at least breaking if you want correct behavior, right); maybe it would be better to rework it so there's an enforced relationship between the old info and the requested name list (eg, by taking the mode in for each name in alongside it, so the index isn't used at all)?

Reply:
I really dont want us to break the API.. So instead of partial resolution info, the resolutionInfo contains the complete information = stringLiterals for modules which can be resolved. So if resolutionInfo is present, we resolved resolutionInfo.names otherwise moduleNames. I think this is fair compromise without breaking the API users.